### PR TITLE
user model revision

### PR DESCRIPTION
**Pull request by johnpaul-tactiv**

### CODE REVIEW

File: users/utils.py [hunk 1]
```diff
@@ -7,3 +7,11 @@ def user_media_path(obj, filename):
 def artwork_media_path(obj, filename):
     """set artwork media path"""
     return f"artwork/{obj.artwork_id}/image/{filename}"
+
+
+def generateHandle(handle, count = 0):
```
Comment (earvingemenez) [inline users/utils.py]: snake case

Comment (johnpaul-tactiv) [reply users/utils.py]: @earvingemenez change the format from the camel case to snake case

File: users/utils.py [hunk 2]
```diff
@@ -7,3 +7,11 @@ def user_media_path(obj, filename):
...
+    if (count):
```
Comment (earvingemenez) [inline users/utils.py]: pwede yata.
```
def generate_handle(handle, count=0):
    return f"{handle}{count and count + 1}"
```

File: users/serializers.py [hunk 3]
```diff
@@ -1,35 +1,48 @@
...
-    def validate_username(self, value):
-        if get_user_model().objects.filter(username=value).exists():
-            raise serializers.ValidationError("Username Already Exist.")
-        return value
     
+    def validate(self, data):
+        handle = data.get('email').split('@')[0]
+        existingHandle = self.Meta.model.objects.filter(handle=handle)
+        generatedHandle = generateHandle(handle, existingHandle.count())
+        data['handle']  = generatedHandle
```
Comment (earvingemenez) [inline users/serializers.py]: use `.update()`

```
data.update({
    'handle': generateHandle(handle, existingHandle.count()),
})

return data
```

File: users/serializers.py [hunk 4]
```diff
@@ -1,35 +1,48 @@
...
+        fields = ('email' ,'first_name', 'last_name', 'password', 'tag')
         extra_kwargs = {'password': {'write_only': True, 'min_length': 5,'style':{'input_type': 'password'}}}
-
-    def validate_username(self, value):
-        if get_user_model().objects.filter(username=value).exists():
-            raise serializers.ValidationError("Username Already Exist.")
-        return value
     
+    def validate(self, data):
+        handle = data.get('email').split('@')[0]
```
Comment (earvingemenez) [inline users/serializers.py]: `handle, domain = data.get('email').split('@')` # avoid indexing error

File: users/serializers.py [hunk 5]
```diff
@@ -1,35 +1,48 @@
...
+
+
+class UserRegistrationSerializer(serializers.ModelSerializer):
+    tag = UserTagSerializer(write_only=True, many=True, allow_null=True, required=False)
 
     class Meta:
         model = get_user_model()
-        fields = ('id','email' ,'first_name', 'last_name', 'description', 'username', 'password','tag')
+        fields = ('email' ,'first_name', 'last_name', 'password', 'tag')
         extra_kwargs = {'password': {'write_only': True, 'min_length': 5,'style':{'input_type': 'password'}}}
```
Comment (earvingemenez) [inline users/serializers.py]: declare it together with `tag` for uniformity and readability

File: users/utils.py (hunk 2 above)
Comment (johnpaul-tactiv) [reply users/utils.py]: ahh okay. now i know.

File: users/serializers.py (hunk 5 above)
Comment (earvingemenez) [reply users/serializers.py]: `password = serializers.CharField(write_only=True, min_length=5)`

File: users/serializers.py [hunk 6]
```diff
@@ -1,35 +1,48 @@
...
+    class Meta:
+        model = UserTag
+        fields = ('tag',)
+
+    def __init__(self, *args, **kwargs):
+        self.owner = kwargs.pop('owner', None)
+        return super(UserTagSerializer, self).__init__(*args, **kwargs)
+
+    def create(self, data):
+        self.Meta.model.objects.create(owner=self.owner ,**data)
```
Comment (earvingemenez) [inline users/serializers.py]: try this:
```
def create(self, data):
   data = super(UserTagSerializer, self).create(owner=self.owner, **data)
   return data
```

File: users/serializers.py [hunk 7]
```diff
@@ -1,35 +1,48 @@
...
-from itertools import islice
-
 from django.contrib.auth import get_user_model, authenticate
 from rest_framework import serializers
 from rest_framework.authtoken.models import Token
 from users.models import UserTag
 
-class UserRegistrationSerializers(serializers.ModelSerializer):
-    tag = serializers.CharField(write_only=True)
+from .utils import generateHandle
```
Comment (earvingemenez) [inline users/serializers.py]: there is no camelCase in python.

functions and variables = snake_case
classes = PascalCase

File: artworks/serializers.py [hunk 8]
```diff
@@ -14,7 +14,7 @@ class Meta:
         fields = ('image',)
 
 class ArtWorkSerializer(serializers.ModelSerializer):
-    owner = UserRegistrationSerializers(read_only=True)
+    owner = UserRegistrationSerializer(read_only=True)
     # image = ArtWorkImageSerializer(many=False, write_only=True, allow_null=True)
```
Comment (earvingemenez) [inline artworks/serializers.py]: remove unneccessary comments

**Comment by earvingemenez** [comment]
@johnpaul-tactiv remove all `.pyc`. please check the .gitignore file why it's not being ignored

File: users/views.py [hunk 9]
```diff
@@ -1,12 +1,24 @@
...
     permission_classes = (permissions.AllowAny,)
-    serializer_class = UserRegistrationSerializers
+    serializer_class = UserRegistrationSerializer
+
+    def create(self, request):
+        serializer = self.serializer_class(
+            data=request.data
+        )
+        
+        if (serializer.is_valid(raise_exception=True)):
```
Comment (earvingemenez) [inline users/views.py]:
```
def create(self, request):
    serializer = self.serializer_class(data=request.data)
    serializer.is_valid(raise_exception=True) # mag return na siya dayun ug error response if invalid
    serializer.save()

    return Response(serializer.data, status=201)
```

ana lang paul para clean and simplier

File: users/utils.py [hunk 10]
```diff
@@ -7,3 +7,7 @@ def user_media_path(obj, filename):
 def artwork_media_path(obj, filename):
     """set artwork media path"""
     return f"artwork/{obj.artwork_id}/image/{filename}"
+
+
+def generate_handle(handle, count = 0):
+    return f"{handle} {(lambda x: x+1 if x else '')(count)}"
```
Comment (earvingemenez) [inline users/utils.py]: naay `space` ang handle ug number? `earvin 1`

Comment (johnpaul-tactiv) [reply users/utils.py]: wala na sya space sir. namali lang wala ko kabantay. na naay space. dapat dikit sila update lang nako.

**Pull request merged**